### ISSUE TITLE
PCA computation and visualization API

### ISSUE THREAD

**Issue by jbeezley** [girder/metabulo#52]
I played around with Lilian's [PCA example](https://app.box.com/file/420100004563).  It provides a number a really good visualization examples and gives me a better idea about how to pass data to the client.

My immediate impressions are that a lot of the computation is hidden in the plotting functions.  If we want to generate the vis client side, then the `prcomp` function is the only part we would be able to use directly, but we could look at the code implementing the plotting functions to extract the relevant portions.

As far as where this leads our API, it depends on the kind of interactivity is desired.  One thing I noticed from the ellipsoids generated in `ggbiplot` is that it computes one ellipsoid per group.  We were originally considering doing the grouping (as in "color by") entirely on the client side.  This would mean either the ellipsoid computation would have to remain client side or we would need to provide a group-by parameter in the API.  I think either approach is reasonable, but doing it server-side will limit some of the interactivity that would be available with a pure client side approach.

We could also take the extreme approach and return the three dataframes (`rotation`, `sdev`, `x`) created by `prcomp` and let the client do *all* of the vis related computation.  Because the PCA-based plots seem to be the primary means by which the data exploration is done, I could see this being the most attractive option as providing the maximum flexibility for interactivity.

### R Script and PCA Details

The content below is primarily to document what I found so I don't forget, but might be interesting for others as well particularly @waxlamp for concrete vis examples.

In this example, I am using `Fol_ACVR2Bfc_muscle_metab_signif_v2.csv` read as follows:
```r
data = read.csv("Fol_ACVR2Bfc_muscle_metab_signif_v2.csv", row.names=1)
df = data[,2:44]
```

Computing the PCA is a one liner (assumes all columns are numeric):
```r
pca <- prcomp(df, rank=max_rank)
```
This returns a data structure containing:
* `pca$rotation`: The eigenvectors (as columns) of the covariance sorted from largest to smallest.
* `pca$sdev`: The square root of the eigenvectors sorted from largest to smallest.
* `pca$x`: The components of the samples projected onto the eigenvectors.  This is what is displayed in our current scatter plot.

The basic grouped scatter plot is generated by:
```r
library(ggfortify)
autoplot(pca, data=data, colour="Musc.Metab")
```
![autoplot](https://user-images.githubusercontent.com/31890/54756326-64e35400-4bbe-11e9-9117-ef929b7c8930.jpeg)

or a default `biplot`:
```r
biplot(pca)
```
![biplot](https://user-images.githubusercontent.com/31890/54756428-96f4b600-4bbe-11e9-829f-3e25af4fabb3.jpeg)

or using [`ggbiplot`](https://github.com/vqv/ggbiplot) described at https://www.datacamp.com/community/tutorials/pca-analysis-r:
```r
library(ggbiplot)
g <- ggbiplot(pca, obs.scale=1, var.scale=1, groups=data[,1], ellipse = TRUE, circle = TRUE)
g <- g + scale_color_discrete(name='')
g <- g + theme(legend.direction='horizontal', legend.position = 'top')
print(g)
```
![ggbiplot](https://user-images.githubusercontent.com/31890/54756592-ed61f480-4bbe-11e9-8da4-530551150e9c.jpeg)

Here is another one using `ggplot2` described at https://www.r-bloggers.com/computing-and-visualizing-pca-in-r/:
```r
library(ggplot2)

theta <- seq(0, 2*pi, length.out = 100)
circle <- data.frame(x=cos(theta), y=sin(theta))
p <- ggplot(circle,aes(x,y)) + geom_path()

loadings <- data.frame(pca$rotation, .names=row.names(pca$rotation))
p + geom_text(data=loadings, mapping=aes(x=PC1, y=PC2, label=.names, colour=.names)) +
  coord_fixed(ratio=1) + labs(x="PC1", y="PC2")
```
![circle](https://user-images.githubusercontent.com/31890/54760080-b8a56b80-4bc5-11e9-9ac0-54e5ed32ad0d.jpeg)

Finally, there is a package called [`NormalizeMets`](https://rdrr.io/cran/NormalizeMets/man/PcaPlots.html) that produces some static html containing interactive plots.
```r
library(NormalizeMets)
PcaPlots(df, data[,1], multiplot=TRUE, varplot=TRUE, interactiveplots=TRUE)
```
The static html seems to be generated using plotly.  I put them up [here](https://jbeezley.github.io/metabolomics-static-plots/loading.html) and [here](https://jbeezley.github.io/metabolomics-static-plots/score.html).  It also includes some other static plots that might be of interest:
![_multiplot](https://user-images.githubusercontent.com/31890/54761337-48e4b000-4bc8-11e9-8041-2b5936655244.png)
![_PCA_loading](https://user-images.githubusercontent.com/31890/54761342-4c783700-4bc8-11e9-82d2-36501bb5ecbc.png)
![_PCA_score](https://user-images.githubusercontent.com/31890/54761562-b264be80-4bc8-11e9-8efb-317bdbc850c3.png)
![_varplot](https://user-images.githubusercontent.com/31890/54761357-526e1800-4bc8-11e9-882e-1607efb8ec1b.png)

**Comment by jbeezley** [comment]
### TLDR

Keep the same API or change the `/csv/:id/pca` endpoint to:

* Compute group level ellipses server side:
`/csv/:id/pca?x=<component #>&y=<component #>&groupby=<metadata column>` =>
```json
{
  "points": [{"x": 1, "y": 2, "group": "value1"}, ...],
  "groups": {
    "group1": {
      "center": {"x": 0.3, "y": 0.5},
      "r1": {"x": 1.5, "y": 0},
      "r2": {"x": 0, "y": 1}
  },
  ...
}
```
* Do all vis-related computations client side:
Return raw PCA output as matrices:
```json
{
  "rotation": [...],
  "sdev": [...],
  "x": [...]
}
```
* Something else?

**Comment by jeffbaumes** [comment]
Given all that information, perhaps sending `rotation` `sdev` `x` is a good idea. Overall it seems like it might be less code to maintain, given that we want interactivity? Do we need to look into a solid stats JS library or should most client-side computation be fairly trivial?

**Comment by waxlamp** [comment]
I would say to return the PCA output from the API, and let the client figure out how to plot the points and draw ellipses, which IIUC should be fairly simple from `rotation` and `sdev`.

One thing I'm not clear on: in the case where you compute one ellipse per group, how is that done? Is it simply repeating the same computations that yield `rotation` and `sdev`, but done once per group? If yes, then it seems like it would be pretty simple to compute those and send them in an auxiliary array of `{rotation: ..., sdev: ...}` objects, labeled by group ID?

**Comment by jbeezley** [comment]
> I would say to return the PCA output from the API, and let the client figure out how to plot the points and draw ellipses, which IIUC should be fairly simple from rotation and sdev.

I could be wrong, but I don't think it can be derived from those matrices.  At least the [R implementation](https://github.com/vqv/ggbiplot/blob/master/R/ggbiplot.r#L185-L201) doesn't.

> One thing I'm not clear on: in the case where you compute one ellipse per group, how is that done? Is it simply repeating the same computations that yield rotation and sdev, but done once per group? If yes, then it seems like it would be pretty simple to compute those and send them in an auxiliary array of {rotation: ..., sdev: ...} objects, labeled by group ID?

They will be calculated from the rows of `x` matching the group label in the same way you have already done with the full sample.  Precomputing them is difficult because it would be a lot ellipsoids to compute and return... `O(number of labels * number of groups per label * number of components^2)`.

> Do we need to look into a solid stats JS library or should most client-side computation be fairly trivial?

Most of the JS libraries I know of are kind of an overkill for this, but maybe there is a specialize library for this purpose (specifically dealing with 2 x 2 covariances).

**Comment by waxlamp** [comment]
>> I would say to return the PCA output from the API, and let the client figure out how to plot the points and draw ellipses, which IIUC should be fairly simple from rotation and sdev.

> I could be wrong, but I don't think it can be derived from those matrices. At least the R implementation doesn't.

Let's clarify this a bit: I thought the data ellipse would simply be the one aligned with the first two eigenvectors (i.e. columns 1 and 2 from `rotation`), scaled by the first two standard deviations (i.e. the first two values of `sdev`). Did I misinterpret the meaning of these two dataframes?

**Comment by waxlamp** [comment]
Just to make it explicit, I'm trying to draw an ellipse that would follow the vectors drawn in cell 6 of this PCA example: https://jakevdp.github.io/PythonDataScienceHandbook/05.09-principal-component-analysis.html

**Comment by jbeezley** [comment]
In that article, there is a two dimensional dataset.  To draw a connection with what we are doing and that, you would have to work in the projected two dimensional space represented by the two principal components in question.  It is mostly likely possible to apply a projection on the rotation matrix to generate the global covariance arrows in this way, but I haven't tried to work it out.

It is more complicated when you want the *per group* ellipses.  That's where I started looking at R code for inspiration, but found it was just doing the naive calculation (direct sample covariance with eigenvector decomposition).  I'm not saying it's not possible to use the rotation matrix, I just don't know for sure unless I spend some time verifying it.

**Comment by jeffbaumes** [comment]
Because of the "M" in "MVP", I suggest punting on most of this for now. Perhaps we don't have a data ellipse, or don't support per-group ellipses yet.

**Comment by waxlamp** [comment]
I actually just had a short conversation with @jbeezley about this, and we think we may have worked out some of the confusion. I have a low-effort vis experiment to try out and if that doesn't show expected results I will punt.

**Comment by waxlamp** [comment]
There's a PCA vis component in there now, but it probably doesn't address all the concerns brought up in this thread. Should we close this issue in light of the PCA component that we have now? If not, we should break the ideas in this thread up into their own issues, maybe.
